src/student_success_tool/analysis/pdp/targets/failure_to_graduate_within_x_years.py

```diff
@@ -0,0 +1,208 @@
+    n: int,
+    enrollment_intensity_col: str = "enrollment_intensity_first_term",
+    term_rank_col: str = "term_rank",
+) -> pd.DataFrame:
+    """
+    Args:
+        df
+        student_criteria
+        student_id_cols
+        intensity_time_lefts
+        max_term_rank
+        num_terms_in_year
+        exclude_pre_cohort_terms
+        term_is_pre_cohort_col
+        n
+        enrollment_intensity_col
+        term_rank_col
+    """
+    nuq_students_in = df.groupby(by=student_id_cols, sort=False).ngroups
+    df_students_by_criteria = shared.select_students_by_criteria(
+        df, student_id_cols=student_id_cols, **student_criteria
+    )
+    if exclude_pre_cohort_terms is True:
+        df_students_by_checkin = shared.get_nth_student_terms_within_cohort(
+            df,
+            student_id_cols=student_id_cols,
+            term_is_pre_cohort_col=term_is_pre_cohort_col,
+            sort_cols=term_rank_col,
+            n=n,
+            include_cols=None,
+        ).loc[:, utils.to_list(student_id_cols)]
+        df_ref = shared.get_first_student_terms_within_cohort(
+            df,
+            student_id_cols=student_id_cols,
+            sort_cols=term_rank_col,
+            include_cols=[enrollment_intensity_col],
+        )
+    else:
+        df_students_by_checkin = shared.get_nth_student_terms(
+            df,
+            student_id_cols=student_id_cols,
+            sort_cols=term_rank_col,
+            n=n,
+            include_cols=None,
+        ).loc[:, utils.to_list(student_id_cols)]
+        df_ref = shared.get_first_student_terms(
+            df,
+            student_id_cols=student_id_cols,
+            sort_cols=term_rank_col,
+            include_cols=[enrollment_intensity_col],
+        )
+    nuq_students_checkin = len(df_students_by_checkin)
+    shared._log_eligible_selection(
+        nuq_students_in, nuq_students_checkin, "check-in point"
+        )
+    df_students_by_time_left = (
+        shared.select_students_by_time_left(
+            df_ref,
+            student_id_cols=student_id_cols,
+            intensity_time_lefts=intensity_time_lefts,
+            max_term_rank=max_term_rank,
+            num_terms_in_year=num_terms_in_year,
+            enrollment_intensity_col=enrollment_intensity_col,
+            term_rank_col=term_rank_col
+        )
+    )
+    df_students_by_time_left = (
+        shared.select_students_by_time_left(
+            df_ref,
+            student_id_cols=student_id_cols,
+            intensity_time_lefts=intensity_time_lefts,
+            max_term_rank=max_term_rank,
+            num_terms_in_year=num_terms_in_year,
+            enrollment_intensity_col=enrollment_intensity_col,
+            term_rank_col=term_rank_col
+        )
+    )
+    df_out = features.shared.merge_many_dataframes(
+        [df_students_by_criteria, df_students_by_checkin, df_students_by_time_left],
+        on=student_id_cols,
+        how="inner",
+    )
```

Comment:
Making sure I understand: You're selecting eligible students by "criteria", by having at least "time left" from first term, and by being enrolled in an nth term? If that's right, could you describe that in the function's docstring?

I recognize that you've used the `failure_to_earn_enough_credits_in_time_from_enrollment` code as your reference example, and that does not have a suitable docstring. Please don't follow that bad example! I was super time-pressed when I wrote that code, and haven't yet gotten round to making it better.